utilities/imgls.pl

```diff
@@ -0,0 +1,559 @@
+#!/usr/bin/perl
+
+# imgls.pl: ls(1) listing supplemented with image thumbnails and dimensions
+#
+#    Usage: imgls [--width #] [--height #] [--[no]preserve_ratio]
+#                 [--[no]dimensions] [--[no]unknown] [ls options] [file ...]
+#
+# Writing an image to an iTerm window is simple. See the official documentation
+# at https://iterm2.com/documentation-images.html and the write_image() function below.
+#
+# Many of ls' options are supported, but not all.  The program does not support ls'
+# default -C columnar output mode - output is always one entry per line.  Writing
+# images across the page in columns appears to be problematic.
+#
+# In addition, options are available to specify setting image properties (width, height,
+# preserve aspect ratio), include inline image dimensions, and disable output of
+# generic icons for unsupported image types. Finally, a table-based image dimensions
+# lookup mechanism is employed to obtain image dimensions.  It can call on Perl
+# modules such as Image::Size, or call on external programs such as sips, mdls or php.
+# It is easy to add additional entries to the table.  You can use the --method option
+# to select any of the currently supported methods ('sips', 'mdls', 'php', and
+# 'image::size').  These are tried, in that order; the first that appears to work
+# is used for all.
+
+use v5.14;
+use strict;
+use utf8;
+use warnings;
+
+use File::stat;
+use File::Which;
+use MIME::Base64;
+use File::Basename;
+use Encode qw(decode);
+use List::Util qw(max);
+use POSIX qw(strftime floor modf);
+use Capture::Tiny ':all';
```

Comment:
What's this for? It isn't on Ubuntu 10 and removing it seems to be harmless.

Reply:
The issue is that the methods to obtain dimensions may write to STDERR, and that now needs to be captured (and not dumped onto the screen).  Unfortunately, Perl makes this hard.  With the old php method, capturing STDOUT (via qx/ /) was sufficient (and Image::Size is not relevant for this matter).  There are over a dozen fork/exex/capture modules in Perl-land, each with their own issues, several of which work by creating temporary files (and as such, require cleanup handlers).

But I'm confused about the Ubuntu reference - is iTerm2/3 coming to Linux-land?

Reply:
You can run imgls anywhere over ssh, although iTerm2 still runs only on macOS.

Would something like this work?
```
use File::Spec;
open STDERR, '>', File::Spec->devnull() or die "could not open STDERR: $!\n";
```

Reply:
Ah, OK.  I was hoping you got your $10 million.

Running this way didn't occurred to me, and now that I think about it, it feels unlikely that an Ubuntu user would have / rely on imgls.pl anyway.  but OK, I'm game.

We have other problems if running imgls.pl on a _stock_ Ubuntu system is a goal.  

On a fresh install, on first run:

`Can't locate File/Which.pm...`

So, File::Which would also have to be either removed from imgls.pl, or installed:

`sudo apt install libfile-which-perl`

Trying to run of course fails due to the missing Capture::Tiny, so a quick install:

`sudo apt install libcapture-tiny-perl`

but woe to me:

```
$ imgls.pl
imgls.pl: no methods found to obtain image dimensions. Tried:
   sips
   mdls
   php
   Image::Size
```

So _something_ has to be installed.  Let's just see if Image::Size gets us further:

`sudo apt install libimage-size-perl`

Yup, now it runs fine from an Ubuntu host over ssh.

Re: your question about the method to redirect STDERR.  I could just ignore STDERR and use qx/ / to capture STDOUT.  But I don't want to just bitbucket STDERR (that's why I'm using this module), as the dimensions methods each have the own issues with certain image formats.  I ran into several, and want the output readily available for development and debugging.  I have in mind supporting multiple methods / invocation, and may rely on STDERR info to help that.

Reply:
I've updated the code.  You should have a pull request awaiting.

- Replaced Capture::Tiny with core modules IO::Open3 and IO::Select (nobody today uses these '90's era, low-level open/select methods, preferring to use safe/sane/tested modules).

- Added support for exiftool method.